File: veilid-flutter/example/integration_test/test_dht.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:veilid/veilid.dart';

final bogusKey =
    RecordKey.fromString('VLD0:qD10lHHPD1_Qr23_Qy-1JnxTht12eaWwENVG_m2v7II');

class TestDHT {
  late final VeilidRoutingContext routingContext;
  final CryptoKind cryptoKind;

  TestDHT(this.cryptoKind);

  Future<void> setUpAll() async {
    routingContext = await Veilid.instance.routingContext();
  }

  Future<void> tearDownAll() async {
    routingContext.close();
  }

  Future<void> testGetDHTValueUnopened() async {
    await expectLater(() => routingContext.getDHTValue(bogusKey, 0),
        throwsA(isA<VeilidAPIException>()));
  }

  Future<void> testOpenDHTRecordNonexistentNoWriter() async {
    await expectLater(() => routingContext.openDHTRecord(bogusKey),
        throwsA(isA<VeilidAPIException>()));
  }

  Future<void> testCloseDHTRecordNonexistent() async {
    await expectLater(() => routingContext.closeDHTRecord(bogusKey),
        throwsA(isA<VeilidAPIException>()));
  }

  Future<void> testDeleteDHTRecordNonexistent() async {
    await expectLater(() => routingContext.deleteDHTRecord(bogusKey),
        throwsA(isA<VeilidAPIException>()));
  }

  Future<void> testCreateDeleteDHTRecordSimple() async {
    final rec = await routingContext.createDHTRecord(
        cryptoKind, const DHTSchema.dflt(oCnt: 1));
    await routingContext.closeDHTRecord(rec.key);
    await routingContext.deleteDHTRecord(rec.key);
  }

  Future<void> testCreateDeleteDHTRecordNoClose() async {
    final rec = await routingContext.createDHTRecord(
        cryptoKind, const DHTSchema.dflt(oCnt: 1));
    await routingContext.deleteDHTRecord(rec.key);
  }

  Future<void> testGetDHTValueNonexistent() async {
    final rec = await routingContext.createDHTRecord(
        cryptoKind, const DHTSchema.dflt(oCnt: 1));
    expect(await routingContext.getDHTValue(rec.key, 0), isNull);
    await routingContext.deleteDHTRecord(rec.key);
  }

  Future<void> testSetGetDHTValue() async {
    final rec = await routingContext.createDHTRecord(
        cryptoKind, const DHTSchema.dflt(oCnt: 2));
    expect(
        await routingContext.setDHTValue(
            rec.key, 0, utf8.encode('BLAH BLAH BLAH')),
        isNull);
    final vd2 = await routingContext.getDHTValue(rec.key, 0);
    expect(vd2, isNotNull);

    final vd3 =
        await routingContext.getDHTValue(rec.key, 0, forceRefresh: true);
    expect(vd3, isNotNull);

    final vd4 = await routingContext.getDHTValue(rec.key, 1);
    expect(vd4, isNull);

    expect(vd2, equals(vd3));

    await routingContext.deleteDHTRecord(rec.key);
  }

  Future<void> testSetGetDHTValueWithOwner() async {
    final cs = await Veilid.instance.getCryptoSystem(cryptoKind);

    final ownerKeyPair = await cs.generateKeyPair();

    final rec = await routingContext.createDHTRecord(
        cryptoKind, const DHTSchema.dflt(oCnt: 2),
        owner: ownerKeyPair);
    expect(
        await routingContext.setDHTValue(
            rec.key, 0, utf8.encode('BLAH BLAH BLAH'),
            options: const SetDHTValueOptions(allowOffline: false)),
        isNull);
    final vd2 = await routingContext.getDHTValue(rec.key, 0);
    expect(vd2, isNotNull);

    final vd3 =
        await routingContext.getDHTValue(rec.key, 0, forceRefresh: true);
    expect(vd3, isNotNull);

    final vd4 = await routingContext.getDHTValue(rec.key, 1);
    expect(vd4, isNull);

    expect(vd2, equals(vd3));

    await routingContext.deleteDHTRecord(rec.key);
  }

  Future<void> testCreateDHTRecordWithDeterministicKey() async {
    final cs = await Veilid.instance.getCryptoSystem(cryptoKind);
    final ownerKeyPair = await cs.generateKeyPair();
    final owner = ownerKeyPair.key;
    final secret = ownerKeyPair.secret;
    const schema = DHTSchema.dflt(oCnt: 1);
    final dhtRecord = await routingContext.createDHTRecord(
        cs.kind(), const DHTSchema.dflt(oCnt: 1),
        owner: ownerKeyPair);
    final encryptionKey = dhtRecord.key.encryptionKey;
    final dhtRecordKey =
        await Veilid.instance.getDHTRecordKey(schema, owner, encryptionKey);
    expect(dhtRecord.key, equals(dhtRecordKey));
    expect(dhtRecord.owner, equals(owner));
    expect(dhtRecord.ownerSecret, equals(secret));
    expect(dhtRecord.schema, equals(schema));
    await routingContext.closeDHTRecord(dhtRecord.key);
    await routingContext.deleteDHTRecord(dhtRecord.key);
  }

  Future<void> testOpenWriterDHTValue() async {
    final cs = await Veilid.instance.getCryptoSystem(cryptoKind);

    var rec = await routingContext.createDHTRecord(
        cs.kind(), const DHTSchema.dflt(oCnt: 2));
    final key = rec.key;
    final owner = rec.owner;
    final secret = rec.ownerSecret!;

    expect(await cs.validateKeyPair(owner, secret), isTrue);
    final otherKeyPair = await cs.generateKeyPair();

    final va = utf8.encode('Qwertyuiop Asdfghjkl Zxcvbnm');
    final vb = utf8.encode('1234567890');
    final vc = utf8.encode(r'!@#$%^&*()');

    // Test subkey writes
    expect(await routingContext.setDHTValue(key, 1, va), isNull);

    var vdtemp = await routingContext.getDHTValue(key, 1);
    expect(vdtemp, isNotNull);
    expect(vdtemp!.data, equals(va));
    expect(vdtemp.seq, equals(0));
    expect(vdtemp.writer, equals(owner));

    expect(await routingContext.getDHTValue(key, 0), isNull);

    expect(await routingContext.setDHTValue(key, 0, vb), isNull);

    expect(
        await routingContext.getDHTValue(key, 0, forceRefresh: true),
        equals(ValueData(
          data: vb,
          seq: 0,
          writer: owner,
        )));

    expect(
        await routingContext.getDHTValue(key, 1, forceRefresh: true),
        equals(ValueData(
          data: va,
          seq: 0,
          writer: owner,
        )));

    // Equal value should not trigger sequence number update
    expect(await routingContext.setDHTValue(key, 1, va), isNull);

    // Different value should trigger sequence number update
    expect(await routingContext.setDHTValue(key, 1, vb), isNull);

    await settle(key, 0);
    await settle(key, 1);

    // Now that we initialized some subkeys
    // and verified they stored correctly
    // Delete things locally and reopen and see if we can write
    // with the same writer key
    //

    await routingContext.closeDHTRecord(key);
    await routingContext.deleteDHTRecord(key);

    rec = await routingContext.openDHTRecord(key,
        writer: KeyPair(key: owner, secret: secret));
    expect(rec, isNotNull);
    expect(rec.key, equals(key));
    expect(rec.owner, equals(owner));
    expect(rec.ownerSecret, equals(secret));
    expect(rec.schema, isA<DHTSchemaDFLT>());
    expect(rec.schema.oCnt, equals(2));

    // Verify subkey 1 can be set before it is get but newer is available online
    vdtemp = await routingContext.setDHTValue(key, 1, vc);
    expect(vdtemp, isNotNull);
    expect(vdtemp!.data, equals(vb));
    expect(vdtemp.seq, equals(1));
    expect(vdtemp.writer, equals(owner));

    // Verify subkey 1 can be set a second time
    // and it updates because seq is newer
    expect(await routingContext.setDHTValue(key, 1, vc), isNull);

    // Verify the network got the subkey update with a refresh check
    vdtemp = await routingContext.getDHTValue(key, 1, forceRefresh: true);
    expect(vdtemp, isNotNull);
    expect(vdtemp!.data, equals(vc));
    expect(vdtemp.seq, equals(2));
    expect(vdtemp.writer, equals(owner));

    // Delete things locally and reopen and see if we can write
    // with a different writer key (should fail)
    await routingContext.closeDHTRecord(key);
    await routingContext.deleteDHTRecord(key);

    rec = await routingContext.openDHTRecord(key, writer: otherKeyPair);
    expect(rec, isNotNull);
    expect(rec.key, equals(key));
    expect(rec.owner, equals(owner));
    expect(rec.ownerSecret, isNull);
    expect(rec.schema, isA<DHTSchemaDFLT>());
    expect(rec.schema.oCnt, equals(2));

    // Verify subkey 1 can NOT be set because we have the wrong writer
    await expectLater(() => routingContext.setDHTValue(key, 1, va),
        throwsA(isA<VeilidAPIException>()));

    // Verify subkey 0 can NOT be set because we have the wrong writer
    await expectLater(() => routingContext.setDHTValue(key, 0, va),
        throwsA(isA<VeilidAPIException>()));

    // Verify subkey 0 can be set because override with the right writer
    // Should have prior sequence number as its returned value because it
    // exists online at seq 0
    vdtemp = await routingContext.setDHTValue(key, 0, va,
        options:
            SetDHTValueOptions(writer: KeyPair(key: owner, secret: secret)));
    expect(vdtemp, isNotNull);
    expect(vdtemp!.data, equals(vb));
    expect(vdtemp.seq, equals(0));
    expect(vdtemp.writer, equals(owner));

    // Should update the second time to seq 1
    vdtemp = await routingContext.setDHTValue(key, 0, va,
        options: SetDHTValueOptions(
            writer: KeyPair(key: owner, secret: secret), allowOffline: false));
    expect(vdtemp, isNull);

    // Clean up
    await routingContext.closeDHTRecord(key);
    await routingContext.deleteDHTRecord(key);
  }

  Future<void> settle(RecordKey key, int subkey) async {
    // Wait for set to settle
    do {
      await Future<void>.delayed(const Duration(milliseconds: 100));
    } while ((await routingContext.inspectDHTRecord(key))
        .offlineSubkeys
        .containsSubkey(subkey));
  }

  Future<void> testInspectDHTRecord() async {
    final rec = await routingContext.createDHTRecord(
        cryptoKind, const DHTSchema.dflt(oCnt: 2));

    expect(
        await routingContext.setDHTValue(
            rec.key, 0, utf8.encode('BLAH BLAH BLAH'),
            options: const SetDHTValueOptions(allowOffline: false)),
        isNull);

    final rr = await routingContext.inspectDHTRecord(rec.key);
    expect(rr.subkeys, equals([ValueSubkeyRange.make(0, 1)]));
    expect(rr.localSeqs, equals([0, null]));
    expect(rr.networkSeqs, equals([null, null]));

    final rr2 = await routingContext.inspectDHTRecord(rec.key,
        scope: DHTReportScope.syncGet);
    expect(rr2.subkeys, equals([ValueSubkeyRange.make(0, 1)]));
    expect(rr2.localSeqs, equals([0, null]));
    expect(rr2.networkSeqs, equals([0, null]));

    await routingContext.closeDHTRecord(rec.key);
    await routingContext.deleteDHTRecord(rec.key);
  }
}

File: veilid-core/src/tests/common/test_dht.rs
use crate::tests_crypto::*;

use super::*;

use lazy_static::*;

lazy_static! {
    static ref BOGUS_KEY: RecordKey = RecordKey::new(
        TEST_CRYPTO_KIND,
        BareRecordKey::new(
            BareOpaqueRecordKey::new(&[0u8; 32]),
            Some(BareSharedSecret::new(&[1u8; 32]))
        )
    );
}

async fn test_get_dht_value_unopened(api: VeilidAPI) {
    let rc = api.routing_context().unwrap();

    let result = rc.get_dht_value(BOGUS_KEY.clone(), 0, false).await;
    assert_err!(result);
}

async fn test_open_dht_record_nonexistent_no_writer(api: VeilidAPI) {
    let rc = api.routing_context().unwrap();

    let result = rc.get_dht_value(BOGUS_KEY.clone(), 0, false).await;
    assert_err!(result);
}

async fn test_close_dht_record_nonexistent(api: VeilidAPI) {
    let rc = api.routing_context().unwrap();

    let result = rc.close_dht_record(BOGUS_KEY.clone()).await;
    assert_err!(result);
}

async fn test_delete_dht_record_nonexistent(api: VeilidAPI) {
    let rc = api.routing_context().unwrap();

    let result = rc.delete_dht_record(BOGUS_KEY.clone()).await;
    assert_err!(result);
}

async fn test_create_delete_dht_record_simple(api: VeilidAPI) {
    let rc = api.routing_context().unwrap();

    let rec = rc
        .create_dht_record(TEST_CRYPTO_KIND, DHTSchema::dflt(1).unwrap(), None)
        .await
        .unwrap();

    let dht_key = rec.key();
    rc.close_dht_record(dht_key.clone()).await.unwrap();
    rc.delete_dht_record(dht_key).await.unwrap();
}

async fn test_create_dht_record_with_owner(api: VeilidAPI) {
    let rc = api.routing_context().unwrap();

    let crypto = api.crypto().unwrap();
    let cs = crypto.get(TEST_CRYPTO_KIND).unwrap();
    let owner_keypair = cs.generate_keypair();

    let rec = rc
        .create_dht_record(
            TEST_CRYPTO_KIND,
            DHTSchema::dflt(1).unwrap(),
            Some(owner_keypair.clone()),
        )
        .await
        .unwrap();

    assert_eq!(rec.ref_owner(), &owner_keypair.key());

    let dht_key = rec.key();
    rc.close_dht_record(dht_key.clone()).await.unwrap();
    rc.delete_dht_record(dht_key.clone()).await.unwrap();
}

async fn test_get_dht_record_key(api: VeilidAPI) {
    let rc = api.routing_context().unwrap();

    let crypto = api.crypto().unwrap();
    let cs = crypto.get(TEST_CRYPTO_KIND).unwrap();
    let owner_keypair = cs.generate_keypair();
    let schema = DHTSchema::dflt(1).unwrap();

    // create the record normally
    let rec = rc
        .create_dht_record(
            TEST_CRYPTO_KIND,
            schema.clone(),
            Some(owner_keypair.clone()),
        )
        .await
        .unwrap();

    // recreate the record key from the metadata alone
    let key = api
        .get_dht_record_key(
            schema.clone(),
            owner_keypair.key(),
            rec.key().encryption_key(),
        )
        .unwrap();

    // keys should be the same
    assert_eq!(key, rec.key());

    let dht_key = rec.key();
    rc.close_dht_record(dht_key.clone()).await.unwrap();
    rc.delete_dht_record(dht_key).await.unwrap();
}

async fn test_get_dht_value_nonexistent(api: VeilidAPI) {
    let rc = api.routing_context().unwrap();

    let rec = rc
        .create_dht_record(TEST_CRYPTO_KIND, DHTSchema::dflt(1).unwrap(), None)
        .await
        .unwrap();
    let dht_key = rec.key();
    let result = rc.get_dht_value(dht_key.clone(), 0, false).await;
    assert_eq!(result.expect("should not be error"), None);

    rc.close_dht_record(dht_key.clone()).await.unwrap();
    rc.delete_dht_record(dht_key).await.unwrap();
}

async fn test_set_get_dht_value(api: VeilidAPI) {
    let rc = api.routing_context().unwrap();

    let rec = rc
        .create_dht_record(TEST_CRYPTO_KIND, DHTSchema::dflt(2).unwrap(), None)
        .await
        .unwrap();
    let dht_key = rec.key();

    let test_value = String::from("BLAH BLAH BLAH").as_bytes().to_vec();
    // convert string to byte array
    let set_dht_value_result = rc
        .set_dht_value(dht_key.clone(), 0, test_value.clone(), None)
        .await;
    assert_eq!(set_dht_value_result.expect("should be Ok(None)"), None);

    let get_dht_value_result_0_non_force = rc.get_dht_value(dht_key.clone(), 0, false).await;
    assert_eq!(
        get_dht_value_result_0_non_force
            .expect("should not be error")
            .expect("should hold a value")
            .data(),
        test_value.clone()
    );

    // works in python, fails in rust
    let get_dht_value_result_0_force = rc.get_dht_value(dht_key.clone(), 0, true).await;
    assert_eq!(
        get_dht_value_result_0_force
            .expect("should not be error")
            .expect("should hold a value")
            .data(),
        test_value.clone()
    );

    let get_dht_value_result_1_non_force = rc.get_dht_value(dht_key.clone(), 1, false).await;
    assert_eq!(
        get_dht_value_result_1_non_force.expect("should not be error"),
        None
    );

    // assert_eq!(
    //     get_dht_value_result_0_non_force.expect("should hold value"),
    //     get_dht_value_result_1_non_force.expect("should hold value")
    // );

    rc.close_dht_record(dht_key.clone()).await.unwrap();
    rc.delete_dht_record(dht_key).await.unwrap();
}

async fn test_open_writer_dht_value(api: VeilidAPI) {
    let rc = api.routing_context().unwrap();

    let rec = rc
        .create_dht_record(TEST_CRYPTO_KIND, DHTSchema::dflt(2).unwrap(), None)
        .await
        .unwrap();
    let key = rec.key();
    let owner = rec.owner();
    let secret = rec.owner_secret().unwrap();
    let keypair = KeyPair::new_from_parts(owner.clone(), secret.value());

    let test_value_1 = String::from("Qwertyuiop Asdfghjkl Zxcvbnm")
        .as_bytes()
        .to_vec();
    let test_data_2 = String::from("1234567890").as_bytes().to_vec();
    let test_data_3 = String::from("!@#$%^&*()").as_bytes().to_vec();

    // Scenario 1
    // 1. Write test data 1 to subkey 1,
    // 2. Read data from subkey 1, without force_refresh, check data, sequence and owner
    // 3. Read data from subkey 0, should return an error
    // 4. Write test data to subkey 0
    // 5. Read data from subkey 0 with force_refresh, check data
    // 6. Read data from subkey 1 with force_refresh, check data
    // 7. Overwrite value 1 twice, check that there's no errors
    let set_dht_test_value_1_result = rc
        .set_dht_value(key.clone(), 1, test_value_1.clone(), None)
        .await;
    assert!(set_dht_test_value_1_result.is_ok());

    let get_dht_value_result_1_non_force = rc.get_dht_value(key.clone(), 1, false).await;
    assert!(get_dht_value_result_1_non_force.is_ok());
    let get_dht_value_result_1_non_force = get_dht_value_result_1_non_force
        .unwrap()
        .expect("should hold value");
    assert_eq!(get_dht_value_result_1_non_force.data(), test_value_1);
    assert_eq!(get_dht_value_result_1_non_force.seq(), 0.into());
    assert_eq!(get_dht_value_result_1_non_force.writer(), owner);

    let get_dht_value_result_0_non_force = rc.get_dht_value(key.clone(), 0, false).await;
    assert_eq!(
        get_dht_value_result_0_non_force.expect("should not be error"),
        None
    );

    let set_dht_test_value_0_result = rc
        .set_dht_value(key.clone(), 0, test_data_2.clone(), None)
        .await;
    assert!(set_dht_test_value_0_result.is_ok());

    let get_dht_value_result_0_force = rc.get_dht_value(key.clone(), 0, true).await;
    assert_eq!(
        get_dht_value_result_0_force
            .expect("should be OK(result)")
            .expect("should hold value")
            .data(),
        test_data_2
    );

    let get_dht_value_result_1_force = rc.get_dht_value(key.clone(), 1, true).await;
    assert_eq!(
        get_dht_value_result_1_force
            .expect("should be OK(result)")
            .expect("should hold value")
            .data(),
        test_value_1
    );

    let overwrite_value_1_result_1 = rc
        .set_dht_value(key.clone(), 1, test_value_1.clone(), None)
        .await;
    assert!(overwrite_value_1_result_1.is_ok());

    let overwrite_value_1_result_2 = rc
        .set_dht_value(key.clone(), 1, test_data_2.clone(), None)
        .await;
    assert!(overwrite_value_1_result_2.is_ok());

    // Now that we initialized some subkeys
    // and verified they stored correctly
    // Delete things locally and reopen and see if we can write
    // with the same writer key

    rc.close_dht_record(key.clone()).await.unwrap();
    rc.delete_dht_record(key.clone()).await.unwrap();

    // Scenario 2
    // 1. Open DHT record with existing keys
    // 2. Check record key, owner, record secret and schema against original values
    // 3. Write test data 3 to subkey 1, without updating a value check that it still
    //    holds test data 2, but sequence has incremented, check owner
    // 4. Check that subkey 1 can be overwritten
    // 5. Read data from subkey 1 with force_refresh, check data

    let rec = rc.open_dht_record(key.clone(), Some(keypair.clone())).await;
    assert!(rec.is_ok());
    let rec = rec.unwrap();
    assert_eq!(rec.ref_key().value(), key.value());
    assert_eq!(rec.ref_key().kind(), key.kind());
    assert_eq!(rec.owner(), owner);
    assert_eq!(rec.owner_secret().unwrap(), secret);
    assert_eq!(rec.ref_schema(), &DHTSchema::dflt(2).unwrap());

    //Verify subkey 1 can be set before it is get but newer is available online
    let set_dht_test_value_1_result = rc
        .set_dht_value(key.clone(), 1, test_data_3.clone(), None)
        .await;
    assert!(set_dht_test_value_1_result.is_ok());
    let vdtemp = set_dht_test_value_1_result.unwrap().unwrap();
    assert_eq!(vdtemp.data(), test_data_2);
    assert_eq!(vdtemp.seq(), 1.into());
    assert_eq!(vdtemp.writer(), owner);

    // Verify subkey 1 can be set a second time and it updates because seq is newer
    let set_dht_test_value_1_result = rc
        .set_dht_value(key.clone(), 1, test_data_3.clone(), None)
        .await;
    assert!(set_dht_test_value_1_result.is_ok());

    // Verify the network got the subkey update with a refresh check
    let get_dht_value_result_1_force = rc.get_dht_value(key.clone(), 1, true).await;
    assert!(get_dht_value_result_1_force.is_ok());
    let get_dht_value_result_1_force = get_dht_value_result_1_force
        .expect("should be OK(result)")
        .expect("should hold value");
    assert_eq!(get_dht_value_result_1_force.data(), test_data_3);
    assert_eq!(get_dht_value_result_1_force.seq(), 2.into());
    assert_eq!(get_dht_value_result_1_force.writer(), owner);

    // Delete things locally and reopen and see if we can write
    // with a different writer key (should fail)
    rc.close_dht_record(key.clone()).await.unwrap();
    rc.delete_dht_record(key.clone()).await.unwrap();

    // Scenario 3
    // 1. Open DHT record with new keypair
    // 2. Check record key, owner, record secret and schema against original values
    // 3. Try writing to subkey 1, expect error
    // 4. Try writing to subkey 0, expect error

    let crypto = api.crypto().unwrap();
    let cs = crypto.get(key.kind()).unwrap();
    assert!(cs
        .validate_keypair(&owner, &secret)
        .expect("should succeed"));
    let other_keypair = cs.generate_keypair();

    let rec = rc.open_dht_record(key.clone(), Some(other_keypair)).await;
    assert!(rec.is_ok());
    let rec = rec.unwrap();
    assert_eq!(rec.ref_key().value(), key.value());
    assert_eq!(rec.ref_key().kind(), key.kind());
    assert_eq!(rec.owner(), owner);
    assert_eq!(rec.owner_secret(), None);
    assert_eq!(rec.ref_schema(), &DHTSchema::dflt(2).unwrap());

    // Verify subkey 1 can NOT be set because we have the wrong writer
    let set_dht_test_value_0_result = rc
        .set_dht_value(key.clone(), 1, test_value_1.clone(), None)
        .await;
    assert_err!(set_dht_test_value_0_result);

    // Verify subkey 0 can NOT be set because we have the wrong writer
    let set_dht_test_value_0_result = rc
        .set_dht_value(key.clone(), 0, test_value_1.clone(), None)
        .await;
    assert_err!(set_dht_test_value_0_result);

    // Verify subkey 0 can be set because we have overridden with the correct writer
    let set_dht_test_value_0_result = rc
        .set_dht_value(
            key.clone(),
            0,
            test_value_1.clone(),
            Some(SetDHTValueOptions {
                writer: Some(keypair),
                allow_offline: None,
            }),
        )
        .await;
    assert!(set_dht_test_value_0_result.is_ok());

    rc.close_dht_record(key.clone()).await.unwrap();
    rc.delete_dht_record(key).await.unwrap();
}

pub async fn test_set_dht_value_allow_offline(api: VeilidAPI) {
    let rc = api.routing_context().unwrap();

    // Create a DHT record
    let rec = rc
        .create_dht_record(CRYPTO_KIND_VLD0, DHTSchema::dflt(1).unwrap(), None)
        .await
        .unwrap();
    let dht_key = rec.key();

    let test_value = String::from("Test offline value").as_bytes().to_vec();

    // Test 1: Default behavior (options = None) should allow offline writes
    let set_result = rc
        .set_dht_value(dht_key.clone(), 0, test_value.clone(), None)
        .await;
    assert!(set_result.is_ok());

    // Test 2: Default behavior (allow_offline = None) should allow offline writes
    let set_result = rc
        .set_dht_value(
            dht_key.clone(),
            0,
            test_value.clone(),
            Some(SetDHTValueOptions {
                writer: None,
                allow_offline: None,
            }),
        )
        .await;
    assert!(set_result.is_ok());

    // Test 3: Explicitly allow offline writes
    let set_result = rc
        .set_dht_value(
            dht_key.clone(),
            1,
            test_value.clone(),
            Some(SetDHTValueOptions {
                writer: None,
                allow_offline: Some(AllowOffline(true)),
            }),
        )
        .await;
    assert!(set_result.is_ok());

    // Test 4: Disallow offline writes
    let set_result = rc
        .set_dht_value(
            dht_key.clone(),
            2,
            test_value.clone(),
            Some(SetDHTValueOptions {
                writer: None,
                allow_offline: Some(AllowOffline(false)),
            }),
        )
        .await;
    assert!(set_result.is_err());
    assert!(set_result.unwrap_err().to_string().contains("offline"));

    rc.close_dht_record(dht_key.clone()).await.unwrap();
    rc.delete_dht_record(dht_key).await.unwrap();
}

pub async fn test_all() {
    let (update_callback, config) = fixture_veilid_core();
    let api = api_startup(update_callback, config)
        .await
        .expect("startup failed");

    let _ = api.attach().await;
    fixture_wait_for_public_internet_ready(&api).await;

    test_get_dht_value_unopened(api.clone()).await;
    test_open_dht_record_nonexistent_no_writer(api.clone()).await;
    test_close_dht_record_nonexistent(api.clone()).await;
    test_delete_dht_record_nonexistent(api.clone()).await;
    test_get_dht_value_nonexistent(api.clone()).await;
    test_create_delete_dht_record_simple(api.clone()).await;
    test_create_dht_record_with_owner(api.clone()).await;
    test_get_dht_record_key(api.clone()).await;
    test_set_get_dht_value(api.clone()).await;
    test_open_writer_dht_value(api.clone()).await;
    test_set_dht_value_allow_offline(api.clone()).await;

    api.shutdown().await;
}
